sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst

```diff
@@ -32,3 +32,184 @@ Arguments:
+   double __devicelib_cbrt(double x);
+   float  __devicelib_cbrtf(float x);
+   double __devicelib_hypot(double x, double y);
+   float  __devicelib_hypotf(float x, float y);
+   double __devicelib_erf(double x);
+   float  __devicelib_erff(float x);
+   double __devicelib_erfc(double x);
+   float  __devicelib_erfcf(float x);
+   double __devicelib_tgamma(double x);
+   float  __devicelib_tgammaf(float x);
+   double __devicelib_lgamma(double x);
+   float  __devicelib_lgammaf(float x);
+   double __devicelib_fmod(double x, double y);
+   float  __devicelib_fmodf(float x, float y);
+   double __devicelib_remainder(double x, double y);
+   float  __devicelib_remainderf(float x, float y);
+   double __devicelib_remquo(double x, double y, int *q);
+   float  __devicelib_remquof(float x, float y, int *q);
+   double __devicelib_nextafter(double x, double y);
+   float  __devicelib_nextafterf(float x, float y);
+   double __devicelib_fdim(double x, double y);
+   float  __devicelib_fdimf(float x, float y);
+   double __devicelib_fma(double x, double y, double z);
+   float  __devicelib_fmaf(float x, float y, float z);
+   float  __devicelib_asinf(float x);
+   double __devicelib_asin(double x);
+   float  __devicelib_atanf(float x);
+   double __devicelib_atan(double x);
+   float  __devicelib_atan2f(float x, float y);
+   double __devicelib_atan2(double x, double y);
+   float  __devicelib_coshf(float x);
+   double __devicelib_cosh(double x);
+   float  __devicelib_sinhf(float x);
+   double __devicelib_sinh(double x);
+   float  __devicelib_tanhf(float x);
+   double __devicelib_tanh(double x);
+   float  __devicelib_acoshf(float x);
+   double __devicelib_acosh(double x);
+   float  __devicelib_asinhf(float x);
+   double __devicelib_asinh(double x);
+   float  __devicelib_atanhf(float x);
+   double __devicelib_atanh(double x);
+   float  __devicelib_frexpf(float x, int *exp);
+   double __devicelib_frexp(double x, int *exp);
+   float  __devicelib_ldexpf(float x, int exp);
+   double __devicelib_ldexp(double x, int exp);
+   float  __devicelib_log10f(float x);
+   double __devicelib_log10(double x);
+   float  __devicelib_modff(float x, float *intpart);
+   double __devicelib_modf(double x, double *intpart);
+   float  __devicelib_expf(float x);
+   double __devicelib_exp(double x);
+   float  __devicelib_exp2f(float x);
+   double __devicelib_exp2(double x);
+   float  __devicelib_expm1f(float x);
+   double __devicelib_expm1(double x);
+   int    __devicelib_ilogbf(float x);
+   int    __devicelib_ilogb(double x);
+   float  __devicelib_log1pf(float x);
+   double __devicelib_log1p(double x);
+   float  __devicelib_log2f(float x);
+   double __devicelib_log2(double x);
+   float  __devicelib_logbf(float x);
+   double __devicelib_logb(double x);
+
+Semantic:
+Those __devicelib_* functions have the same semantic as corresponding math functions
+from <math.h> or <cmath>. Please refer to: http://www.cplusplus.com/reference/cmath/
+for details.
```

Comment:
> The fallback implementation for C math functions are based on `__spirv_ocl*` math built-ins.

AFAIK, semantics of `__spirv_ocl*` math built-ins deviates from C math library semantics (see section 7 of OpenCL specification).
It means that the same function will behave differently depending on whether native or fallback function implementation is used. How user is supposed to reason about function behavior?

Reply:
This is a good question. In fact, all math functions supported currently are wrappers of corresponding sycl ocl math builtins. And sycl::* math functions are wrappers of corresponding __spirv_ocl_* math builtin too, can we say: Those __devicelib_* functions have the same semantic as corresponding sycl math functions? And we can ask users to refer to sycl spec.
Thank you very much.

Reply:
I think it should be the opposite. Fallback device implementation of C/C++ math functions must follow C/C++ math functions semantics on the host to be consistent. 
SYCL math functions semantics is different and currently aligned with `__spirv_ocl*` functions.

Reply:
When you write a kernel, the semantics should be the same regardless of where the kernel is executed. I don't see how we can promise C/C++ math function semantics because we can't, in general, do things like setting errno.

Reply:
> I don't see how we can promise C/C++ math function semantics because we can't, in general, do things like setting errno.

I think all deviations from C/C++ math function semantics must be documented.
@andykaylor, do you have a full list of things which can't be supported on the accelerators?

Reply:
Hi, Alexey and Andy. I also think we need to document the deviations from C/C++ and I need to spend some time looking up the Linux manual and SYCL spec to find those deviations. Currently, the first things comes to my mind is some math functions are supposed to set errno and float-point exception(do we support fp-exeception in SYCL kernel?).
And we have "cl::sycl::rint" in SYCL math functions, the corresponding math functions in Linux is rint and following description is from Linux man page:

"The  rint(),  rintf(),  and rintl() functions do the same, but will raise the inexact exception (FE_INEXACT, checkable via fetestexcept(3)) when the result differs in value from the argument."
I looked into SYCL spec and didn't claim cl::sycl::rint would raise exception or not. Does sycl or OpenCL spec has clear defined whether such FP related exception is supported?
Thank you very much.

Reply:
I think most of the differences are covered in section 7 of OpenCL specification.

Reply:
To functions in current math device libraries, I searched the libm document and found most of those functions may set errno and raise FP-exceptions and I look up the OCL spec and found FP-exception is disabled in OCL. We need document that "errno and float-point exception are not supported in SYCL device library".

Reply:
@bader As far as I know the complete list of things that can't be supported is errno and FP exception handling. The standard requires an identifier, math_errhandling, which indicates what form of error handling is supported. If we do not set the MATH_ERRNO bit in this value, the math functions are not expected to correctly set errno. If we do not set MATH_ERREXCEPT, the math functions are not expected to raise FP exceptions. Strictly speaking, the standard requires conforming implementations to set one or both of these values, but I think it we set neither and document this as a known divergence from the standard our behavior should be well understood.

The OpenCL specification defines some requirements for boundary condition results that go beyond the C99 requirements. I assume the built-in implementations must already be meeting these requirements for all devices we support.

Reply:
Already modified the DeviceLibExtensions.rst according to Andy's suggestion.